core/src/main/java/hudson/security/BasicAuthenticationFilter.java

```diff
@@ -163,7 +164,7 @@ public void doFilter(ServletRequest request, ServletResponse response, FilterCha
 
         // prepare a redirect
         rsp.setStatus(HttpServletResponse.SC_MOVED_TEMPORARILY);
-        rsp.setHeader("Location",path);
+        rsp.setHeader("Location", URLEncoder.encode(path, StandardCharsets.UTF_8.displayName()));
```

Comment:
It should be fine so, according to https://docs.oracle.com/javaee/6/api/javax/servlet/http/HttpServletResponse.html#setHeader(java.lang.String,%20java.lang.String) but as many others said, better to have some tests.